### ISSUE TITLE
Why is there a problem when loading saved weights on a model

### ISSUE THREAD

**Issue by Atralb** [keras-team/keras#12094]
Hi everybody, 

I'm trying to modify a classifier model with many tools (dropout, autoencoder, etc...) to analyse what gets the best results. Thus, I am using the `save_weights` and `load_weights` methods. 

The first time I am launching my model, it works fine. However when loading the weights, the `fit` isn't doing anything. The loss stagnates during the entire training. 

I thought first that it was an issue of vanishing gradient since I encountered the problem first with the autoencoded dataset. But after many tweaks and tries, I feel the issue resides in the weights loading. See for yourself (this is of course done after a Runtime Restart) : 

```
# Classifier

model = Sequential()

model.add(Dense(50, activation= 'relu', input_dim= x.shape[1]))
model.add(Dense(50, activation= 'relu'))
model.add(Dense(50, activation= 'relu'))
model.add(Dense(50, activation= 'relu'))
model.add(Dense(10, activation= 'softmax'))

model.compile(optimizer='adam', loss = 'categorical_crossentropy', metrics = ['acc'])

model.save_weights('/content/drive/My Drive/Colab Notebooks/Weights/KagTPOneStart')
```
### First time training
loading initial weight then `fit`. (Yes I know the initial weights are already there at this time but I left the line here voluntarily to prove that the first time loading the weights doesn't cause a problem): 

```
model.load_weights('/content/drive/My Drive/Colab Notebooks/Weights/KagTPOneStart')

model.fit(x,y_train,epochs=10,batch_size=20, validation_split=0.15)

model.save_weights('/content/drive/My Drive/Colab Notebooks/Weights/KagTPOneNormal')
```
Results : 

```
Train on 35700 samples, validate on 6300 samples
Epoch 1/10
35700/35700 [==============================] - 5s 128us/step - loss: 1.0875 - acc: 0.8036 - val_loss: 0.3275 - val_acc: 0.9067
Epoch 2/10
35700/35700 [==============================] - 4s 120us/step - loss: 0.2792 - acc: 0.9201 - val_loss: 0.3186 - val_acc: 0.9079
Epoch 3/10
35700/35700 [==============================] - 4s 122us/step - loss: 0.2255 - acc: 0.9357 - val_loss: 0.1918 - val_acc: 0.9444
Epoch 4/10
35700/35700 [==============================] - 4s 121us/step - loss: 0.1777 - acc: 0.9499 - val_loss: 0.1977 - val_acc: 0.9465
Epoch 5/10
35700/35700 [==============================] - 4s 121us/step - loss: 0.1530 - acc: 0.9549 - val_loss: 0.1718 - val_acc: 0.9478
Epoch 6/10
35700/35700 [==============================] - 4s 121us/step - loss: 0.1402 - acc: 0.9595 - val_loss: 0.1847 - val_acc: 0.9510
Epoch 7/10
35700/35700 [==============================] - 4s 122us/step - loss: 0.1236 - acc: 0.9637 - val_loss: 0.1675 - val_acc: 0.9546
Epoch 8/10
35700/35700 [==============================] - 4s 121us/step - loss: 0.1160 - acc: 0.9660 - val_loss: 0.1776 - val_acc: 0.9586
Epoch 9/10
35700/35700 [==============================] - 4s 120us/step - loss: 0.1109 - acc: 0.9683 - val_loss: 0.1928 - val_acc: 0.9492
Epoch 10/10
35700/35700 [==============================] - 4s 120us/step - loss: 0.1040 - acc: 0.9701 - val_loss: 0.1749 - val_acc: 0.9570
WARNING:tensorflow:This model was compiled with a Keras optimizer (<tensorflow.python.keras.optimizers.Adam object at 0x7fb76ca35080>) but is being saved in TensorFlow format with `save_weights`. The model's weights will be saved, but unlike with TensorFlow optimizers in the TensorFlow format the optimizer's state will not be saved.

Consider using a TensorFlow optimizer from `tf.train`.
```
### Second time training 
loading initial weights then fit : 

```
model.load_weights('/content/drive/My Drive/Colab Notebooks/Weights/KagTPOneStart')

model.fit(x,y_train,epochs=10,batch_size=20, validation_split=0.15)

model.save_weights('/content/drive/My Drive/Colab Notebooks/Weights/KagTPOneNormal')
```
Results : 

```
Train on 35700 samples, validate on 6300 samples
Epoch 1/10
35700/35700 [==============================] - 4s 121us/step - loss: 14.4847 - acc: 0.1011 - val_loss: 14.5907 - val_acc: 0.0948
Epoch 2/10
35700/35700 [==============================] - 4s 122us/step - loss: 14.5018 - acc: 0.1003 - val_loss: 14.5907 - val_acc: 0.0948
Epoch 3/10
35700/35700 [==============================] - 4s 120us/step - loss: 14.5018 - acc: 0.1003 - val_loss: 14.5907 - val_acc: 0.0948
Epoch 4/10
35700/35700 [==============================] - 4s 121us/step - loss: 14.5018 - acc: 0.1003 - val_loss: 14.5907 - val_acc: 0.0948
Epoch 5/10
35700/35700 [==============================] - 4s 121us/step - loss: 14.5018 - acc: 0.1003 - val_loss: 14.5907 - val_acc: 0.0948
Epoch 6/10
35700/35700 [==============================] - 4s 121us/step - loss: 14.5018 - acc: 0.1003 - val_loss: 14.5907 - val_acc: 0.0948
Epoch 7/10
35700/35700 [==============================] - 4s 122us/step - loss: 14.5018 - acc: 0.1003 - val_loss: 14.5907 - val_acc: 0.0948
Epoch 8/10
35700/35700 [==============================] - 4s 121us/step - loss: 14.5018 - acc: 0.1003 - val_loss: 14.5907 - val_acc: 0.0948
Epoch 9/10
35700/35700 [==============================] - 4s 122us/step - loss: 14.5018 - acc: 0.1003 - val_loss: 14.5907 - val_acc: 0.0948
Epoch 10/10
35700/35700 [==============================] - 5s 130us/step - loss: 14.5018 - acc: 0.1003 - val_loss: 14.5907 - val_acc: 0.0948
WARNING:tensorflow:This model was compiled with a Keras optimizer (<tensorflow.python.keras.optimizers.Adam object at 0x7fb76ca35080>) but is being saved in TensorFlow format with `save_weights`. The model's weights will be saved, but unlike with TensorFlow optimizers in the TensorFlow format the optimizer's state will not be saved.

Consider using a TensorFlow optimizer from `tf.train`.
```
Thanks in advance for your help :)

PS : Here's the data for reference, but I really don't think this is the problem. This is a MNIST-like dataset provided by google on kaggle. (I believe it is exactly MNIST but not all samples) : 

```
import pandas as pd 
df=pd.read_csv('/content/drive/My Drive/Colab Notebooks/IA/Kaggle TP1/train.csv')
data = df.values
data.shape
y = data[:,0]
y_train =  np_utils.to_categorical(y, 10)
x = data[:,1:]
```

**Comment by Atralb** [comment]
@ParikhKadam @msymp Maybe you could help me ?

**Comment by colt18** [comment]
I'm not sure but it may be related to this:
https://stackoverflow.com/questions/47266383/save-and-load-weights-in-keras
Saving weights and saving the whole training process at current epoch are different things. Can you try to `model.save(filepath)` instead of `model.save_weights('my_model_weights.h5')`. If you are resuming the training you need to save all the info.

**Comment by msymp** [comment]
@Atralb , did you consider getting rid of the WARNING, regarding the incompatibility of the model formats, being compiled as Keras optimizer format and saved as TensorFlow format? 
Also let us know if @colt18 suggestion was useful. Thanks.

**Comment by ParikhKadam** [comment]
@Atralb Sorry for being late.. I was busy with issues in my project and today when I saw this thread, it seems just the same like mine. I too faced a problem with resuming training. In my case, the weights are not saved properly (according to me). You should check that out too!

**Comment by DeepBlender** [comment]
This is the exact same issue I am encountering as well in TensorFlow 2.0.

I am using:
model.save_weights(model_directory, save_format='tf')
and
model.load_weights(model_directory)

It clearly loads weights as the result is very different compared to starting from scratch, but it is way off and not usable for the continuation of the training.

Let me know if you need additional information or another reproduction. It would likely take some time to create another reproduction, but if you think it helps, I would gladly provide it.

**Comment by colt18** [comment]
@DeepBlender Try to use model.save instead of model.save_weights. model.save_weights saves only weights of the current model. If you resume training afterwards model.load_weights all other params will init from scratch like the current learning rate, step size etc etc. model.save builds up a save file that includes all the data so when you deploy with model.load and continue, it will work smoothly.

**Comment by DeepBlender** [comment]
@colt18 Thanks for the suggestion. My goal is to save subclassed models and reuse and combine them however I want. That's why save_weights is the easiest solution for me.
According to the documentation save_format='tf' should not only save the weights, but also the optimizer's state. It doesn't matter whether I am using 'tf' or 'h5', in both cases, it is not able to recover and pick up the training. It quickly diverges and is not anymore able to recover.

**Comment by DeepBlender** [comment]
I figure a few things out regarding my use case. When I train my model for 10 epochs, it converges pretty continuously. If I stop after 5 and continue, it doesn't improve or even diverges. That's why I am pretty confident, there is something wrong at least in the TensorFlow 2.0 implementation.
However, I wasn't able to reproduce this with a smaller example which I don't understand at all!
When the training is stopped after 5 epochs and continued with a lower learning rate, it successfully improves. It is highly likely that I am dealing with a too fragile architecture. But it still doesn't explain why the training can't be continued.

**Comment by jvishnuvardhan** [comment]
Please fill the issue [template](https://github.com/keras-team/keras/issues/new?template=a--tensorflow-backend-users.md). Could you update them if they are relevant in your case, or leave them as N/A? Thanks!

**Comment by DeepBlender** [comment]
As the issue can't be reliably reproduced, I fear it would be a waste of time.
I am still trying to understand it better to create an appropriate reproduction. Once I achieved that, I will create another issue or provide the information here with the template.

**Comment by OmriPi** [comment]
Hi, jumping in to add that I've encountered this exact problem.
I am training an NLP model using Bert from tensorflow Hub.
I train the model for 20 epochs and get fantastic results on my dataset.
If I then evaluate it immediately on the test set (no saving and loading), I get really good results on the test set (around 0.97 F1 score).
However, if I save the model, stop the run, load the model and try evaluation on the test set again (without training as it shouldn't be required), I get around 0.017 F1. The obvious conclusion is that something is very wrong with either the saving or the loading of the model. I save the entire model using the ModelCheckpoint callback, and load it using load_model(). I'm at a complete loss as to why this is happening, and would be very glad if anyone could help, as this seems like quite a serious issue...

Thanks!

**Comment by jvishnuvardhan** [comment]
@OmriPi Please fill the issue template [here](https://github.com/keras-team/keras/issues/new/choose). Could you update them if they are relevant in your case, or leave them as N/A? Along with the template, please provide as many details as possible to find the root cause of the issue. It would be great if you can provide a small code to reproduce the error. Thanks!

**Comment by jvishnuvardhan** [comment]
Closing due to lack of recent activity. Please update the issue when new information becomes available, and we will open a new issue. Thanks!

**Comment by DeepBlender** [comment]
When I made my experiments, I was forced to use Google Colab due to a broken computer. As I continued the training, the results got quite funky in plenty of cases. Not only in the negative way, but right now I also have a "too good to be true" result.
I have rerun plenty of those experiments now on my computer and I am not able to reproduce the weirdness at all and it works exactly as expected!

It seems very likely that the issue is caused by something on Google Colab and is not a general problem in TensorFlow (2.0) or Keras.

**Comment by naomifridman** [comment]
Same problem on Google colaboratory.
I saved weights to Google Drive, I see the file there.
But model.load_weights('full path to the file on the drive')
not working. the model behave as a fresh one.

**Comment by jvishnuvardhan** [comment]
@naomifridman Can you post a standalone code here and also post issue in colab repo [here](https://github.com/googlecolab/colabtools/issues). Thanks!

**Comment by karmatta** [comment]
This is not a standalone Google colaboratory issue. I am facing the same issue outside of google colab and the loaded model seems to have re-initialized weights after saving.

I am actually surprised that this has remained open for so long. Has none of you here deployed a model in production? if yes, how did you save the model/weights in order for you to get consistent predictions (as during training)?

**Comment by alvaroslm** [comment]
I think I'm having the same issue. I had some old code and models from last June, using TF1. I updated the code to work in TF2 (mostly just chanigng tf.python.keras to tf.keras), and it can load OLD models written with tf1 and it seems to train ok in TF2, BUT if you save the model now and try to load it, it's not working properly any more. All the predictions are the same no matter what input you provide, it's as if the weights were all zeros (but not the biases). The results changes every time I train and save and load and test, but it's the same for all input images.

**Comment by RhythmParadox** [comment]
I too have noticed my models not loading or saving weights properly.

**Comment by MernaMustafa** [comment]
I'm having the same issue

**Comment by lostdatum** [comment]
Same problem here using tf.keras.callbacks.ModelCheckpoint and tf.keras.models.load_model.
Using Tensorflow 2.0 in Google Colab (saving to mounted Drive).
Using the default ModelCheckpoint params => calls mymodel.save() with new .pb format (not H5).

This is very, very bad.
I will try saving weights only to check whether the problem occurs at saving or loading.

**Comment by yuh8** [comment]
This has worked for me for saving and loading weights

- use `model.to_json()` to save your architecture config only without weights, and training configuration
- save weights using `model.save_weights()`
- load model using `model_new = tf.keras.models.model_from_json(model_json)`
- compile model using `model_new.compile()`
- load weights using `model_new.load_weights()`
 After doing above, I was able to get the same level of accuracy
